lib/Segments/DynamicSegments/Filters/WooCommerceCategory.php

```diff
@@ -30,44 +33,126 @@ public function __construct(
   public function apply(QueryBuilder $queryBuilder, DynamicSegmentFilterEntity $filter): QueryBuilder {
     global $wpdb;
     $filterData = $filter->getFilterData();
-    $categoryId = (int)$filterData->getParam('category_id');
+
+    $operator = $filterData->getOperator();
+    $categoryIds = (array)$filterData->getParam('category_ids');
+    $categoryIdswithChildrenIds = $this->getCategoriesWithChildren($categoryIds);
+
     $subscribersTable = $this->entityManager->getClassMetadata(SubscriberEntity::class)->getTableName();
+
+    $parameterSuffix = $filter->getId() ?: Security::generateRandomString();
+    $parameterSuffix = (string)$parameterSuffix;
+
+    $completedOrder = "postmeta.post_id NOT IN ( SELECT id FROM {$wpdb->posts} AS p WHERE p.post_status IN ('wc-cancelled', 'wc-failed'))";
+
+    if ($operator === DynamicSegmentFilterData::OPERATOR_ANY) {
+      $this->applyPostmetaJoin($queryBuilder, $subscribersTable);
+      $this->applyOrderItemsJoin($queryBuilder);
+      $this->applyOrderItemmetaJoin($queryBuilder);
+      $this->applyTermRelationshipsJoin($queryBuilder);
+      $this->applyTermTaxonomyJoin($queryBuilder, $parameterSuffix);
+
+    } elseif ($operator === DynamicSegmentFilterData::OPERATOR_ALL) {
+      $this->applyPostmetaJoin($queryBuilder, $subscribersTable);
+      $this->applyOrderItemsJoin($queryBuilder);
+      $this->applyOrderItemmetaJoin($queryBuilder);
+      $this->applyTermRelationshipsJoin($queryBuilder);
+      $this->applyTermTaxonomyJoin($queryBuilder, $parameterSuffix)
+      ->groupBy("{$subscribersTable}.id, items.order_id")
+      ->having('COUNT(items.order_id) = :count')
```

Comment:
@triple0t  Sorry for the late comment. Each parameter name needs to be a unique per filter. When we merge queries for multiple filters we also [need to merge the parameters](https://github.com/mailpoet/mailpoet/blob/master/lib/Segments/DynamicSegments/FilterHandler.php#L50-L59) and if the name is not unique the parameters override. So in this case, if somebody combines 2 these segments with `all` operand and selects different counts of categories only one count will be applied for both subqueries in the combined query.

I've just noticed the same issue in the purchased product segment which is already merged. I'll prepare a fix for the products segment. Could you please fix this PR. You just need to use 

```php
->having('COUNT(items.order_id) = :count' . $parameterSuffix)
->setParameter('count' . $parameterSuffix, count($categoryIds));
```

Reply:
@costasovo Noted 👍